serverless/serverless
# Add plugin hooks to define config variable getters

## What did you implement:

added options for plugins to define their own getters for variables



## How did you implement it:

* refactored `getValueFromSource` to iterate through an array of regex & func name pairs
* `PluginManager` now looks for a `variableResovers` entry on plugin instances and if present, adds them to the list of regexes & getters to check in `getValueFromSource`
	* `variableResovers` is an object of the shape:
      ```javascript
      {
        test: getTestVariable, // shortform. ${test:foobar} implemented by getTestVariable
        test2: {
          resolver: getTestVariable,
          name: 'test', // needed if below is true
          isDisabledAtPrepopulation: true // disable when populating stage/region/creds/etc
        }
      }
      ```
(the last object with `dependendServiceName` is optional)


## How can we verify it:
* install this branch
* cd to a serverless project or create a new one
* create a `.serverless_plugins` directory if it does not exists
* create `.serverless_plugins/test.js` containing:
  ```javascript
  class TestPlugin {
    constructor(sls) {

      const getTestVariable = async () => 'testVariable';
      this.variableResolvers = {test: getTestVariable};
    }
  }
  
  module.exports = TestPlugin;
  ```
* add `test` to the list of plugins in `serverless.yml`
* add a value somewhere in `serverless.yml` that uses our variable syntax: `${test:something}`
* run `sls print` and the test variable should now be resolved to `'testVariable'`


## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

       **Validate via `npm test`**
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_

**_Is this ready for review?:_** YES  
**_Is it a breaking change?:_** NO
